src/Collapse/include/gudhi/Flag_complex_edge_collapser.h

```diff
@@ -15,366 +16,309 @@
-    sparse_row_adjacency_matrix_[rw_u].insert(rw_v) = idx;
-    sparse_row_adjacency_matrix_[rw_v].insert(rw_u) = idx;
-    return std::minmax(rw_u, rw_v);
+    return true;
+#else
+    auto&&nc = neighbors[c];
+    // if few neighbors, use dichotomy? Seems slower.
+    // I tried storing a copy of neighbors as a vector<absl::flat_hash_map> and using it for nc, but it was
+    // a bit slower here. It did help with neighbors[dominator].find(w) in the main function though,
+    // sometimes enough, sometimes not.
+    auto ci = nc.begin();
+    auto ce = nc.end();
+    auto eni = e_ngb.begin();
+    auto ene = e_ngb.end();
+    assert(eni != ene);
+    assert(ci != ce);
+    // if(*eni == c && ++eni == ene) return true;
+    for(;;){
+      Vertex ve = *eni;
+      Vertex vc = ci->first;
+      while(ve > vc) {
+        // try a gallop strategy (exponential search)? Seems slower
+        if(++ci == ce) return false;
+        vc = ci->first;
+      }
+      if(ve < vc) return false;
+      // ve == vc
+      if(ci->second > f) return false;
+      if(++eni == ene)return true;
+      // If we stored an open neighborhood of c (excluding c), we would need to test for c here and before the loop
+      // if(*eni == c && ++eni == ene)return true;
+      if(++ci == ce) return false;
+    }
+#endif
   }
 
- public:
-  /** \brief Flag_complex_edge_collapser constructor from a range of filtered edges.
-   *
-   * @param[in] edges Range of Filtered edges range.There is no need the range to be sorted, as it will be performed in
-   * `Flag_complex_edge_collapser::process_edges`.
-   *
-   * \tparam FilteredEdgeRange must be a range for which std::begin and std::end return iterators on a
-   * `Flag_complex_edge_collapser::Filtered_edge`.
-   */
-  template<typename FilteredEdgeRange>
-  Flag_complex_edge_collapser(const FilteredEdgeRange& edges)
-  : f_edge_vector_(std::begin(edges), std::end(edges)) { }
-
-  /** \brief Performs edge collapse in a increasing sequence of the filtration value.
-   *
-   * \tparam filtered_edge_output is a functor that is called on the output edges, in non-decreasing order of
-   * filtration, as filtered_edge_output(u, v, f) where u and v are Vertex_handle representing the extremities of the
-   * edge, and f is its new Filtration_value.
-   */
-  template<typename FilteredEdgeOutput>
-  void process_edges(FilteredEdgeOutput filtered_edge_output) {
-    // Sort edges
-    auto sort_by_filtration = [](const Filtered_edge& edge_a, const Filtered_edge& edge_b) -> bool
+  template<class FilteredEdgeRange, class Delay>
+  void process_edges(FilteredEdgeRange const& edges, Delay&& delay) {
     {
-      return (std::get<2>(edge_a) < std::get<2>(edge_b)); 
-    };
+      Vertex maxi = 0, maxj = 0;
+      for(auto& fe : edges) {
+        Vertex i = std::get<0>(fe);
+        Vertex j = std::get<1>(fe);
+        if (i > maxi) maxi = i;
+        if (j > maxj) maxj = j;
+      }
+      num_vertices = std::max(maxi, maxj) + 1;
```

Comment:
That would just overwrite the value with the one for the last edge :confused: 
I want the maximum of all the vertices in all the edges.
I could maintain just one max instead of 2, having 2 makes it more likely the compiler can use SIMD, but I didn't check if it actually does (`maxi=std::max(maxi,i)` would actually be better than `if(i>maxi)maxi=i` for that), and it is unlikely to take a significant amount of time.

Reply:
Oops sorry, I meant:
```cpp
num_vertices = std::max(num_vertices, std::max(std::get<0>(fe), std::get<1>(fe)) + 1);
```

Reply:
Then the answer is that my version is faster :wink:
(possibly relevant: https://gcc.gnu.org/bugzilla/show_bug.cgi?id=105053 https://gcc.gnu.org/bugzilla/show_bug.cgi?id=105062)

Reply:
Ok, I close this point